File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

isDiv :: Int -> Int -> Bool
isDiv x y = (mod x y) == 0

factors :: Int -> [Int]
factors n = filter (isDiv n) [1,2..(n-1)]

classify :: Int -> Maybe Classification
classify n
  | n < 1 = Nothing
  | s == n = Just Perfect
  | s < n = Just Deficient
  | s > n = Just Abundant
  | otherwise = error "Not possible!"
  where s = sum . factors $ n

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError("Classification is only possible for positive integers.")
    aliquot = sum(i for i in range(1, number) if number % i == 0)
    if aliquot == number:
        return "perfect"
    elif aliquot > number:
        return "abundant"
    else:
        return "deficient"


# vim:ts=2:sw=2:expandtab
